File: 09_d1_discounts/python/main.py
import sys
from enum import Enum


class Day(Enum):
    MONDAY = 'MONDAY',
    TUESDAY = 'TUESDAY'
    WEDNESDAY = 'WEDNESDAY',
    THURSDAY = 'THURSDAY',
    FRIDAY = 'FRIDAY',
    SATURDAY = 'SATURDAY',
    SUNDAY = 'SUNDAY'


def get_purchase_total():
    while True:
        try:
            total = float(input('Ingrese el valor de la compra del cliente > '))
            if total <= 0:
                print('El valor de la compra debe ser un número mayor a cero\n', file=sys.stderr)
                continue
            break
        except:
            print('El valor de la compra debe ser un número entero o decimal válido', file=sys.stderr)

    return total


def get_purchase_day():
    print("\nLunes (L)\nMartes (M)\nMiercoles(W)\nJueves (J)\nViernes (V)\nSabado (S)\nDomingo (D)")

    while True:
        try:
            match input('Ingrese la letra que corresponde al día de la compra según la lista mostrada > ').lower():
                case 'l':
                    day = Day.MONDAY
                case 'm':
                    day = Day.TUESDAY
                case 'w':
                    day = Day.WEDNESDAY
                case 'j':
                    day = Day.THURSDAY
                case 'v':
                    day = Day.FRIDAY
                case 's':
                    day = Day.SATURDAY
                case 'd':
                    day = Day.SUNDAY
                case _:
                    raise Exception('Día inválido. Únicamente se aceptan: L|M|W|J|V|S|D\n')
            break
        except Exception as err:
            print(err, file=sys.stderr)

    return day


def calculate_discount(total, day):
    if total <= 100_000:
        return 0

    achieve_the_highest_discount = total >= 200_000

    match day:
        case Day.MONDAY | Day.WEDNESDAY | Day.FRIDAY:
            discount_percentage = 0.15 if achieve_the_highest_discount else 0.1
        case Day.TUESDAY | Day.THURSDAY:
            discount_percentage = 0.18 if achieve_the_highest_discount else 0.12
        case Day.SATURDAY | Day.SUNDAY:
            discount_percentage = 0.2 if achieve_the_highest_discount else 0.15

    return total * discount_percentage


def main():
    print("\n-- PROGRAMA PARA CALCULAR DE DESCUENTOS POR COMPRAS - ALMACENES D1 --\n")

    purchase_total = get_purchase_total()
    purchase_day = get_purchase_day()
    purchase_discount = calculate_discount(purchase_total, purchase_day)

    if purchase_discount > 0:
        print(f"\nEl descuento para la compra es de: ${purchase_discount:.2}")
    else:
        print("\nLa compra no tiene descuento")


if __name__ == '__main__':
    main()

File: 09_d1_discounts/rust/src/main.rs
use std::io;

enum Day {
    Monday,
    Tuesday,
    Wednesday,
    Thursday,
    Friday,
    Saturday,
    Sunday,
}

fn main() {
    println!("\n-- PROGRAMA PARA CALCULAR DE DESCUENTOS POR COMPRAS - ALMACENES D1 --\n");

    let purchase_total = get_purchase_total();
    let purchase_day = get_purchase_day();
    let purchase_discount = calculate_discount(purchase_total, purchase_day);

    if purchase_discount > 0.0 {
        println!("\nEl descuento para la compra es de: $ {purchase_discount:.2}");
    } else {
        println!("\nLa compra no tiene descuento");
    }
}

fn get_purchase_total() -> f32 {
    let mut input = String::new();

    loop {
        println!("Ingrese el valor de la compra del cliente:");
        io::stdin().read_line(&mut input).unwrap();

        match input.trim().parse::<f32>() {
            Err(_) => {
                eprintln!("El valor de la compra debe ser un número entero o decimal válido\n")
            }
            Ok(total) if total < 0_f32 => {
                eprintln!("El valor de la compra debe ser un número positivo mayor a cero\n")
            }
            Ok(total) => break total,
        }

        input.clear();
    }
}

fn get_purchase_day() -> Day {
    let mut input = String::new();

    println!(
        "\nLunes (L)\nMartes (M)\nMiercoles(W)\nJueves (J)\nViernes (V)\nSabado (S)\nDomingo (D)"
    );

    loop {
        println!("Ingrese la letra que corresponde al día de la compra según la lista mostrada:");
        io::stdin().read_line(&mut input).unwrap();

        break match input.trim().to_lowercase().as_str() {
            "l" => Day::Monday,
            "m" => Day::Tuesday,
            "w" => Day::Wednesday,
            "j" => Day::Thursday,
            "v" => Day::Friday,
            "s" => Day::Saturday,
            "d" => Day::Sunday,
            _ => {
                eprintln!("Día inválido. Únicamente se aceptan: L|M|W|J|V|S|D\n");
                input.clear();

                continue;
            }
        };
    }
}

fn calculate_discount(total: f32, day: Day) -> f32 {
    if total <= 100000_f32 {
        return 0.0;
    }

    let achieve_the_highest_discount = total >= 200000_f32;

    let discount_percentage: f32 = match day {
        Day::Monday | Day::Wednesday | Day::Friday => {
            if achieve_the_highest_discount {
                0.15
            } else {
                0.1
            }
        }
        Day::Tuesday | Day::Thursday => {
            if achieve_the_highest_discount {
                0.18
            } else {
                0.12
            }
        }
        Day::Saturday | Day::Sunday => {
            if achieve_the_highest_discount {
                0.2
            } else {
                0.15
            }
        }
    };

    discount_percentage * total
}
